# Car Analysis

Analyze somthing about German eBay car sales?

### General Project Setup

In [200]:
import numpy as np
import pandas as pd


In [201]:
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [202]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [203]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [204]:
autos.head()


dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

In [205]:
autos["notRepairedDamage"].unique()

array(['nein', nan, 'ja'], dtype=object)

### Initial Observations

1. 5 columns contain null value
    - vehicleType
    - gearbox
    - model
    - fuelType
    - notRepairedDamage
2. Price column is a string but should be int or float.  Will need to strip $ and comma and convert it
3. odometer column should be an int - will need to strip off km and convert it
4. notRepairedDamage has values of 'nein', nan, 'ja' that will need to be decifered in addition to null values
5. column names are camel case and we want them to have underscores


    

In [206]:
columns_before = autos.columns

In [207]:
autos.rename(columns = {'dateCrawled': 'date_crawled', 
                        'name': 'name', 
                        'seller': 'seller', 
                        'offerType': 'offer_type', 
                        'price': 'price', 
                        'abtest': 'abtest',
                        'vehicleType': 'vehicle_type', 
                        'yearOfRegistration': 'registration_year',
                        'gearbox': 'gearbox', 
                        'powerPS': 'power_ps', 
                        'model': 'model',
                        'odometer': 'odometer', 
                        'monthOfRegistration': 'registration_month', 
                        'fuelType': 'fuel_type', 
                        'brand': 'brand',
                        'notRepairedDamage': 'unrepaired_damage', 
                        'dateCreated': 'ad_created', 
                        'nrOfPictures': 'nbr_of_pictures', 
                        'postalCode': 'postal_code',
                        'lastSeen': 'last_seen'},
             inplace=True)
columns_after = autos.columns

print(columns_before)
print(columns_after)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')
Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nbr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


Columns have been renamed to made them both more readable and more standard.

In [208]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nbr_of_pictures   postal_code  \
count                 50000          50000.0  50000.000000   
unique                   76              NaN           NaN   
top     2016-04-03 00:00:00              NaN           NaN   
freq                   1946              NaN           NaN   
mean                    NaN              0.0  50813.627300   
std                     NaN              0.0  25779.747957   
min                     NaN              0.0   1067.000000   
25%                     NaN              0.0  30451.000000   
50%                     NaN              0.0  49577.000000   
75%                     NaN              0.0  71540.000000   
max                     NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### More Observations and required fixes

1. Columns that can be safely removed
    - nbr_of_pictures - all values are 0
    - seller - this column has a single non unique value (gewerblich).  If we remove the corresponding row, we can then remove the entire column since it would have no unique values.
    - offer type - this column has a single non unique value (Gesuch).  If we remove the corresponding row, we can then remove the entire column since it would have no unique values.
2.  Columns that need to be converted to numeric type
    - odometer
    - price

In [209]:
autos['seller'].value_counts(dropna=False)


privat        49999
gewerblich        1
Name: seller, dtype: int64

### Convert String Columns to Int

In [210]:
autos['price'].value_counts()

$0         1421
$500        781
$1,500      734
$2,500      643
$1,000      639
           ... 
$15,823       1
$3,125        1
$29,600       1
$1,770        1
$14,899       1
Name: price, Length: 2357, dtype: int64

In [211]:
if autos['price'].dtype == object:
        autos['price'] = autos['price'].str.replace("$", "").str.replace(",","").astype(int)
autos['price'].value_counts()

0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64

In [212]:
if autos['odometer'].dtype == object:
    autos['odometer'] = autos['odometer'].str.replace("km", "").str.replace(",","").astype(int)
autos['odometer'].value_counts()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer, dtype: int64

In [213]:
autos = autos.rename(columns = {'odometer': 'odometer_km'})
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                 50000        50000   50000      50000  5.000000e+04   
unique                48213        38754       2          2           NaN   
top     2016-03-21 20:37:19  Ford_Fiesta  privat    Angebot           NaN   
freq                      3           78   49999      49999           NaN   
mean                    NaN          NaN     NaN        NaN  9.840044e+03   
std                     NaN          NaN     NaN        NaN  4.811044e+05   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.100000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  1.000000e+08   

       abtest vehicle_type  registration_year  gearbox      power_ps  model  \
count   50000        44905       50000.000000    47320  50000.000000  47242   
unique      2            8                NaN        2           NaN    245   
top      test    limousine                NaN  manuell           NaN   golf   
freq    25756        12859                NaN    36993           NaN   4024   
mean      NaN          NaN        2005.073280      NaN    116.355920    NaN   
std       NaN          NaN         105.712813      NaN    209.216627    NaN   
min       NaN          NaN        1000.000000      NaN      0.000000    NaN   
25%       NaN          NaN        1999.000000      NaN     70.000000    NaN   
50%       NaN          NaN        2003.000000      NaN    105.000000    NaN   
75%       NaN          NaN        2008.000000      NaN    150.000000    NaN   
max       NaN          NaN        9999.000000      NaN  17700.000000    NaN   

          odometer_km  registration_month fuel_type       brand  \
count    50000.000000        50000.000000     45518       50000   
unique            NaN                 NaN         7          40   
top               NaN                 NaN    benzin  volkswagen   
freq              NaN                 NaN     30107       10687   
mean    125732.700000            5.723360       NaN         NaN   
std      40042.211706            3.711984       NaN         NaN   
min       5000.000000            0.000000       NaN         NaN   
25%     125000.000000            3.000000       NaN         NaN   
50%     150000.000000            6.000000       NaN         NaN   
75%     150000.000000            9.000000       NaN         NaN   
max     150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nbr_of_pictures   postal_code  \
count              40171                50000          50000.0  50000.000000   
unique                 2                   76              NaN           NaN   
top                 nein  2016-04-03 00:00:00              NaN           NaN   
freq               35232                 1946              NaN           NaN   
mean                 NaN                  NaN              0.0  50813.627300   
std                  NaN                  NaN              0.0  25779.747957   
min                  NaN                  NaN              0.0   1067.000000   
25%                  NaN                  NaN              0.0  30451.000000   
50%                  NaN                  NaN              0.0  49577.000000   
75%                  NaN                  NaN              0.0  71540.000000   
max                  NaN                  NaN              0.0  99998.000000   

                  last_seen  
count                 50000  
unique                39481  
top     2016-04-07 06:17:27  
freq                      8  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Dig a bit deeper to clean numeric columns

In [214]:
def explore_data(series):
    print('****Series Shape****')
    print(series.unique().shape)
    print('****Series Desciption****')
    print(series.describe())
    print('****Series Unique Values****')
    print(series.unique())
    print('****Series Counts****')
    print(series.value_counts())
    print('****Series Top 5****')
    print(series.value_counts().sort_index(ascending=False).head())
    print('****Series Bottom 5****')
    print(series.value_counts().sort_index(ascending=True).head())

In [215]:
explore_data(autos['price'])
autos.sort_values(['price'], ascending=False)

****Series Shape****
(2357,)
****Series Desciption****
count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64
****Series Unique Values****
[ 5000  8500  8990 ...   385 22200 16995]
****Series Counts****
0        1421
500       781
1500      734
2500      643
1000      639
         ... 
20790       1
8970        1
846         1
2895        1
33980       1
Name: price, Length: 2357, dtype: int64
****Series Top 5****
99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price, dtype: int64
****Series Bottom 5****
0    1421
1     156
2       3
3       1
5       2
Name: price, dtype: int64


date_crawled                                               name  \
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
47598  2016-03-31 18:56:54  Opel_Vectra_B_1_6i_16V_Facelift_Tuning_Showcar...   
...                    ...                                                ...   
43510  2016-03-20 21:50:00                                Verkauf_oder_Tausch   
33620  2016-03-26 18:39:00  VW_Jetta_Oldtimer_Youngtimer_Sammler_Retro_1.8...   
39231  2016-03-29 08:57:50                             Suchen_7_Sitzer_//_Bus   
48060  2016-04-02 11:54:30                            Bmw_Verkauf_oder_Tausch   
38354  2016-03-23 04:36:59               Verkaufe_auto_bekommt_kein_tuev_mehr   

       seller offer_type     price   abtest vehicle_type  registration_year  \
39705  privat    Angebot  99999999  control    limousine               1999   
42221  privat    Angebot  27322222  control    limousine               2014   
27371  privat    Angebot  12345678  control          NaN               2017   
39377  privat    Angebot  12345678  control          NaN               2018   
47598  privat    Angebot  12345678  control    limousine               2001   
...       ...        ...       ...      ...          ...                ...   
43510  privat    Angebot         0     test   kleinwagen               1998   
33620  privat    Angebot         0     test          NaN               2018   
39231  privat    Angebot         0  control          bus               2005   
48060  privat    Angebot         0     test        kombi               1999   
38354  privat    Angebot         0     test          NaN               2000   

         gearbox  power_ps     model  odometer_km  registration_month  \
39705  automatik       224  s_klasse       150000                   9   
42221    manuell       163        c4        40000                   2   
27371        NaN        95     punto       150000                   0   
39377    manuell        95       v40       150000                   6   
47598    manuell       101    vectra       150000                   3   
...          ...       ...       ...          ...                 ...   
43510    manuell       102  a_klasse       150000                   1   
33620  automatik        90       NaN       150000                   3   
39231        NaN         0       NaN        80000                   0   
48060    manuell       150       5er       150000                   3   
38354        NaN         0       NaN       150000                   0   

      fuel_type           brand unrepaired_damage           ad_created  \
39705    benzin   mercedes_benz               NaN  2016-03-22 00:00:00   
42221    diesel         citroen               NaN  2016-03-08 00:00:00   
27371       NaN            fiat               NaN  2016-03-09 00:00:00   
39377       NaN           volvo              nein  2016-03-08 00:00:00   
47598    benzin            opel              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
43510    benzin   mercedes_benz              nein  2016-03-20 00:00:00   
33620       NaN      volkswagen              nein  2016-03-26 00:00:00   
39231       NaN  sonstige_autos              nein  2016-03-29 00:00:00   
48060    benzin             bmw              nein  2016-04-02 00:00:00   
38354       NaN           mazda               NaN  2016-03-23 00:00:00   

       nbr_of_pictures  postal_code            last_seen  
39705                0        73525  2016-04-06 05:15:30  
42221                0        76532  2016-03-08 20:39:05  
27371                0        96110  2016-03-09 15:45:47  
39377                0        14542  2016-04-06 23:17:31  
47598                0         

In [216]:
high = autos[autos['price'] > 900000]
print(high)

              date_crawled                                               name  \
514    2016-03-17 09:53:08                   Ford_Focus_Turnier_1.6_16V_Style   
2897   2016-03-12 21:50:57   Escort_MK_1_Hundeknochen_zum_umbauen_auf_RS_2000   
7814   2016-04-04 11:53:31                                        Ferrari_F40   
11137  2016-03-29 23:52:57  suche_maserati_3200_gt_Zustand_unwichtig_laufe...   
22947  2016-03-22 12:54:19                         Bmw_530d_zum_ausschlachten   
24384  2016-03-21 13:57:51                            Schlachte_Golf_3_gt_tdi   
27371  2016-03-09 15:45:47                                         Fiat_Punto   
37585  2016-03-29 11:38:54                                Volkswagen_Jetta_GT   
39377  2016-03-08 23:53:51                        Tausche_volvo_v40_gegen_van   
39705  2016-03-22 14:58:27                        Tausch_gegen_gleichwertiges   
42221  2016-03-08 20:39:05                                  Leasinguebernahme   
43049  2016-03-21 19:53:52  

In [217]:
low = autos[autos['price'] > 10000000]
low.describe()


price  registration_year    power_ps    odometer_km  \
count  7.000000e+00           7.000000    7.000000       7.000000   
mean   2.665450e+07        2002.428571  103.714286  120000.000000   
std    3.285683e+07          15.935883   73.033587   51316.014394   
min    1.111111e+07        1973.000000    0.000000   40000.000000   
25%    1.172839e+07        1997.000000   71.500000  100000.000000   
50%    1.234568e+07        2001.000000   95.000000  150000.000000   
75%    1.983395e+07        2015.500000  132.000000  150000.000000   
max    1.000000e+08        2018.000000  224.000000  150000.000000   

       registration_month  nbr_of_pictures   postal_code  
count            7.000000              7.0      7.000000  
mean             3.285714              0.0  54007.571429  
std              3.251373              0.0  39958.875046  
min              0.000000              0.0   4356.000000  
25%              1.000000              0.0  16530.500000  
50%              3.000000              0.0  73525.000000  
75%              4.500000              0.0  85500.500000  
max              9.000000              0.0  96110.000000

#### Price Observations

- There are 1421 rows with a cost of $0.  Anything with a cost less than $100 should be removed (1762 rows) as it should be considered bad data.
- There are rows with a price value of $12345678 which is just someone typing in random numbers, and it is too high of a value for a car
- In fact anything above $10,000,000 should be removed (7 rows)

In [218]:
price_adjusted_autos = autos[autos["price"].between(100,10000000)]
cleaned_data = price_adjusted_autos
original_shape = autos.shape
new_shape = price_adjusted_autos.shape
print(original_shape)
print(new_shape)

(50000, 20)
(48231, 20)


In [219]:
explore_data(cleaned_data['odometer_km'])

****Series Shape****
(13,)
****Series Desciption****
count     48231.000000
mean     125916.008376
std       39546.446538
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64
****Series Unique Values****
[150000  70000  50000  80000  10000  30000 125000  90000  20000  60000
   5000 100000  40000]
****Series Counts****
150000    31215
125000     5038
100000     2102
90000      1733
80000      1412
70000      1214
60000      1153
50000      1010
40000       814
30000       777
5000        761
20000       757
10000       245
Name: odometer_km, dtype: int64
****Series Top 5****
150000    31215
125000     5038
100000     2102
90000      1733
80000      1412
Name: odometer_km, dtype: int64
****Series Bottom 5****
5000     761
10000    245
20000    757
30000    777
40000    814
Name: odometer_km, dtype: int64


#### Odometer_km. observations

This data looks pretty clean with no real outliers.

### Explore Date Fields

In [220]:
def explore_date_as_string(series):
    '''
    series expected to me in 'yyyy-dd-mm hh:mm:ss' format
    '''
    dates = series.str[:10]
    print(dates.unique().shape)
#     print(dates.value_counts().sort_index(ascending=True))
    print(dates.value_counts(normalize=True, dropna=False).sort_index())
    

In [221]:
explore_date_as_string(cleaned_data['date_crawled'])

(34,)
2016-03-05    0.025357
2016-03-06    0.014037
2016-03-07    0.036056
2016-03-08    0.033174
2016-03-09    0.033008
2016-03-10    0.032282
2016-03-11    0.032593
2016-03-12    0.036906
2016-03-13    0.015675
2016-03-14    0.036657
2016-03-15    0.034314
2016-03-16    0.029462
2016-03-17    0.031515
2016-03-18    0.012896
2016-03-19    0.034729
2016-03-20    0.037797
2016-03-21    0.037217
2016-03-22    0.032904
2016-03-23    0.032282
2016-03-24    0.029442
2016-03-25    0.031494
2016-03-26    0.032303
2016-03-27    0.031121
2016-03-28    0.034957
2016-03-29    0.034148
2016-03-30    0.033733
2016-03-31    0.031847
2016-04-01    0.033692
2016-04-02    0.035600
2016-04-03    0.038606
2016-04-04    0.036574
2016-04-05    0.013062
2016-04-06    0.003172
2016-04-07    0.001389
Name: date_crawled, dtype: float64


date_crawled has a somewhat even spread across an acceptable date range.

In [222]:
explore_date_as_string(cleaned_data['ad_created'])

(76,)
2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036926
2016-04-05    0.011797
2016-04-06    0.003255
2016-04-07    0.001244
Name: ad_created, Length: 76, dtype: float64


ad_created has an aceptable date range but a larger skew than some of the other date columns.  This makes sense since it is the day the ad was created for the car to be sold, not the date the crawler did anything so you shoudl expect it to be a much larger spread.

In [223]:
explore_date_as_string(cleaned_data['last_seen'])

(34,)
2016-03-05    0.001078
2016-03-06    0.004313
2016-03-07    0.005432
2016-03-08    0.007319
2016-03-09    0.009579
2016-03-10    0.010636
2016-03-11    0.012399
2016-03-12    0.023781
2016-03-13    0.008874
2016-03-14    0.012627
2016-03-15    0.015861
2016-03-16    0.016442
2016-03-17    0.028094
2016-03-18    0.007319
2016-03-19    0.015758
2016-03-20    0.020651
2016-03-21    0.020547
2016-03-22    0.021356
2016-03-23    0.018577
2016-03-24    0.019759
2016-03-25    0.019096
2016-03-26    0.016670
2016-03-27    0.015550
2016-03-28    0.020858
2016-03-29    0.022309
2016-03-30    0.024694
2016-03-31    0.023823
2016-04-01    0.022848
2016-04-02    0.024901
2016-04-03    0.025129
2016-04-04    0.024528
2016-04-05    0.125086
2016-04-06    0.221973
2016-04-07    0.132135
Name: last_seen, dtype: float64


last_seen seems to have valid dates and the spread is not skewed too much.

In [224]:
cleaned_data['registration_year'].describe()

count    48231.000000
mean      2004.728867
std         87.891383
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### 'registration_year' observations

The max of 9999 for registration year is bad data and would need to be examined and possibly removed.  The same is true of the minimum year of 1000.  To give it the most inclusive filter, we could only allow registration dates bewtween 1885 (year car was invented) and the 2016 (year the data is from).

In [225]:
registration_adjusted_autos = cleaned_data[cleaned_data["registration_year"].between(1885,2016)]
cleaned_data = registration_adjusted_autos
print(registration_adjusted_autos['registration_year'].describe())
registration_adjusted_autos['registration_year'].value_counts(normalize=True, dropna=False).sort_index()

count    46359.000000
mean      2002.937876
std          7.130887
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64


1910    0.000043
1927    0.000022
1929    0.000022
1931    0.000022
1934    0.000043
          ...   
2012    0.028215
2013    0.017278
2014    0.014280
2015    0.008197
2016    0.025928
Name: registration_year, Length: 78, dtype: float64

'registration_year' now has a much more expected spread with dates from the early 1900s not have many cars registered while more cars are registered more recently.

### Aggregate by Car Brand

In [229]:
cleaned_data['brand'].value_counts(normalize=True, dropna=False)

volkswagen        0.211415
bmw               0.110184
opel              0.107228
mercedes_benz     0.096637
audi              0.086758
ford              0.069846
renault           0.047067
peugeot           0.029854
fiat              0.025605
seat              0.018249
skoda             0.016415
nissan            0.015337
mazda             0.015229
smart             0.014194
citroen           0.014043
toyota            0.012791
hyundai           0.010009
sonstige_autos    0.009599
volvo             0.009124
mini              0.008801
mitsubishi        0.008175
honda             0.007873
kia               0.007075
alfa_romeo        0.006665
porsche           0.006018
suzuki            0.005932
chevrolet         0.005673
chrysler          0.003516
dacia             0.002653
daihatsu          0.002502
jeep              0.002287
subaru            0.002114
land_rover        0.002114
saab              0.001661
jaguar            0.001532
daewoo            0.001488
trabant           0.001359
r

I'm going to aggregate on any brands with more than 1% of the listings.

In [273]:
brand_list = cleaned_data['brand'].value_counts(normalize=True, dropna=False).sort_index()
partial_brand_list = brand_list[brand_list >= .01]
print(partial_brand_list)
mean_price_dict = {}
mean_mileage_dict = {}

for brand in partial_brand_list.index:
    rows = cleaned_data.loc[cleaned_data['brand'] == brand]
    mean_price_dict[brand] = rows['price'].mean()
    mean_mileage_dict[brand] = rows['odometer_km'].mean()

audi             0.086758
bmw              0.110184
citroen          0.014043
fiat             0.025605
ford             0.069846
hyundai          0.010009
mazda            0.015229
mercedes_benz    0.096637
nissan           0.015337
opel             0.107228
peugeot          0.029854
renault          0.047067
seat             0.018249
skoda            0.016415
smart            0.014194
toyota           0.012791
volkswagen       0.211415
Name: brand, dtype: float64


Prices of the cars with more than 1% of the listings range from about $2,500 (renault) to $9,380(audi).

In [284]:
price_series = pd.Series(mean_price_dict)
km_series = pd.Series(mean_mileage_dict)

price_mileage_df = pd.DataFrame(price_series, columns=['mean_price'])
price_mileage_df['mean_km'] = km_series

price_mileage_df.sort_values(['mean_price'], ascending=False)

mean_price        mean_km
audi           9380.718548  129245.400298
mercedes_benz  8672.654241  131025.669643
bmw            8621.729052  132698.707909
skoda          6409.609724  110906.701708
volkswagen     5639.900316  128804.203653
hyundai        5411.075431  106885.775862
toyota         5167.091062  115944.350759
nissan         4756.659634  118326.300985
seat           4433.419622  121536.643026
mazda          4129.774788  124553.824363
ford           4086.934219  124277.331686
citroen        3796.262673  119554.531490
smart          3596.402736   99734.042553
peugeot        3113.860549  127127.890173
opel           3005.496077  129384.429692
fiat           2836.873631  116950.294861
renault        2496.070577  128281.393217

There is not a correlation between the mean price and the mean odometer reading.